src/Api/QueryParameterValidator/QueryParameterValidator.php

```diff
@@ -72,4 +80,76 @@ public function validateFilters(string $resourceClass, array $resourceFilters, a
             throw new FilterValidationException(array_merge(...$errorList));
         }
     }
+
+    /**
+     * @param array<string, array<string, mixed>> $filterDescription
+     */
+    private static function getCollectionFormat(array $filterDescription): string
+    {
+        return $filterDescription['openapi']['collectionFormat'] ?? $filterDescription['swagger']['collectionFormat'] ?? 'csv';
+    }
+
+    /**
+     * @param array<string, mixed> $queryParameters
+     *
+     * @throws \InvalidArgumentException
+     */
+    private static function iterateValue(string $name, array $queryParameters, string $collectionFormat = 'csv'): \Generator
+    {
+        $candidates = array_filter(
+            $queryParameters,
+            static fn (string $key) => $key === $name || "{$key}[]" === $name,
+            \ARRAY_FILTER_USE_KEY
+        );
+
+        foreach ($candidates as $key => $value) {
+            $values = self::getValue($value, $collectionFormat);
+            foreach ($values as $v) {
+                yield [$key => $v];
+            }
+        }
+    }
+
+    /**
+     * @param int|int[]|string|string[] $value
+     *
+     * @return int[]|string[]
+     */
+    private static function getValue(int|string|array $value, string $collectionFormat = 'csv'): array
+    {
+        if (\is_array($value)) {
+            return $value;
+        }
+
+        if (\is_string($value)) {
+            return explode(self::getSeparator($collectionFormat), $value);
+        }
+
+        return [$value];
+    }
+
+    /** @return non-empty-string */
+    private static function getSeparator(string $collectionFormat): string
+    {
+        return match ($collectionFormat) {
+            'csv' => ',',
+            'ssv' => ' ',
+            'tsv' => '\t',
+            'pipes' => '|',
+            default => throw new \InvalidArgumentException(sprintf('Unknown collection format %s', $collectionFormat)),
+        };
+    }
+
+    private function validate(string $name, array $data, array $queryParameters): array
+    {
+        $errorList = [];
+
+        foreach ($this->validators as $validator) {
+            if ($errors = $validator->validate($name, $data, $queryParameters)) {
+                $errorList[] = $errors;
+            }
+        }
+
+        return array_merge(...$errorList);
```

Comment:
why the array_merge?

Reply:
It was an attempt to reduce the nesting of array leading to the previous `Array to string conversion`. It fixes unit tests but seems to break elsewhere. 

I started to add more tests but it's not yet completed (hence the "draft" state)

Reply:
@soyuka most of the tests are passing now, I need to find a way to avoid thoses `array_merge(...$var)` used to flatten errors lists. I think `iterable` might be the solution. WDT ?